Non-deterministic test failure in pec

Should be seeded most likely. Perhaps increasing num_samples from 100 to 500 would also be good. What do you think @andreamari?

```
================================== FAILURES ===================================
___________ test_execute_with_pec_with_different_samples[circuit1] ____________

circuit = cirq.Circuit([
    cirq.Moment(
        cirq.Y(cirq.LineQubit(1)),
    ),
    cirq.Moment(
        cirq.CNOT(cirq.LineQubit(0), cirq.LineQubit(1)),
    ),
    cirq.Moment(
        cirq.Y(cirq.LineQubit(1)),
    ),
])

    @mark.parametrize("circuit", [oneq_circ, twoq_circ])
    def test_execute_with_pec_with_different_samples(circuit: Circuit):
        """Tests that, on average, the error decreases as the number of samples is
        increased.
        """
        errors_few_samples = []
        errors_more_samples = []
        for _ in range(10):
            mitigated = execute_with_pec(
                circuit, executor, decomposition_dict=DECO_DICT, num_samples=10
            )
            errors_few_samples.append(abs(mitigated - 1.0))
            mitigated = execute_with_pec(
                circuit, executor, decomposition_dict=DECO_DICT, num_samples=100
            )
            errors_more_samples.append(abs(mitigated - 1.0))
    
>       assert np.average(errors_more_samples) < np.average(errors_few_samples)
E       assert 0.05254450883168323 < 0.051046586922745815
E        +  where 0.05254450883168323 = <function average at 0x000001D90B444AF0>([0.0480414521858088, 0.07978086396775974, 0.1009197741092831, 0.015803496651313753, 0.02552260104182469, 0.11349465243866219, ...])
E        +    where <function average at 0x000001D90B444AF0> = np.average
E        +  and   0.051046586922745815 = <function average at 0x000001D90B444AF0>([0.029537088949564194, 0.0267618070541239, 0.0846984748302384, 0.0846984748302384, 0.029537088949564194, 0.029537088949564194, ...])
E        +    where <function average at 0x000001D90B444AF0> = np.average

mitiq\pec\tests\test_pec.py:109: AssertionError
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,9 @@
 
 ### All Changes
 
+- Add refernce to review paper in docs (@willzeng, gh-423).
+- Add unitary folding API (@rmlarose, gh-429).
+
 ## Version 0.3.0 (October 30th, 2020)
 
 ### Summary

--- a/docs/source/apidoc.rst
+++ b/docs/source/apidoc.rst
@@ -97,5 +97,5 @@ Inference and Extrapolation: Factories
 
 Noise Scaling: Unitary Folding
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-.. automodule:: mitiq.zne.scaling
+.. automodule:: mitiq.zne.scaling.folding
    :members:

--- a/mitiq/pec/pec.py
+++ b/mitiq/pec/pec.py
@@ -27,6 +27,7 @@ def execute_with_pec(
     executor: Callable[[QPROGRAM], float],
     decomposition_dict: DecompositionDict,
     num_samples: Optional[int] = None,
+    random_state: Optional[Union[int, np.random.RandomState]] = None,
     full_output: bool = False,
 ) -> Union[float, Tuple[float, float]]:
     """Evaluates the expectation value associated to the input circuit
@@ -53,6 +54,7 @@ def execute_with_pec(
             of the quasi-probability representation of the input circuit.
             Note: the latter feature is not yet implemented and num_samples
             is just set to 1000 if not specified.
+        random_state: Seed for sampling circuits.
         full_output: If False only the average PEC value is returned.
             If True an estimate of the associated error is returned too.
 
@@ -91,7 +93,7 @@ def execute_with_pec(
     for _ in range(num_samples):
         # Note: the norm is the same for each sample.
         sampled_circuit, sign, norm = sample_circuit(
-            circuit, decomposition_dict
+            circuit, decomposition_dict, random_state
         )
         sampled_circuits.append(sampled_circuit)
         signs.append(sign)

--- a/mitiq/pec/sampling.py
+++ b/mitiq/pec/sampling.py
@@ -15,7 +15,7 @@
 
 """Tools for sampling from the noisy decomposition of ideal operations."""
 
-from typing import Tuple, List
+from typing import List, Optional, Tuple, Union
 from copy import deepcopy
 import numpy as np
 
@@ -29,7 +29,9 @@ from mitiq.pec.utils import (
 
 
 def sample_sequence(
-    ideal_operation: Operation, decomposition_dict: DecompositionDict
+    ideal_operation: Operation,
+    decomposition_dict: DecompositionDict,
+    random_state: Optional[Union[int, np.random.RandomState]] = None,
 ) -> Tuple[List[Operation], int, float]:
     """Samples an implementable sequence from the PEC decomposition of the
     input ideal operation. Moreover it also returns the "sign" and "norm"
@@ -40,6 +42,7 @@ def sample_sequence(
             sequence is sampled.
         decomposition_dict: The decomposition dictionary from which the
             decomposition of the input ideal_operation can be extracted.
+        random_state: Seed for sampling.
 
     Returns:
         imp_seq: The sampled implementable sequence as list of one
@@ -47,12 +50,25 @@ def sample_sequence(
         sign: The sign associated to sampled sequence.
         norm: The one norm of the decomposition coefficients.
     """
+    if not random_state:
+        rng = np.random
+    else:
+        if isinstance(random_state, int):
+            rng = np.random.RandomState(random_state)
+        elif isinstance(random_state, np.random.RandomState):
+            rng = random_state
+        else:
+            raise ValueError(
+                "Bad type for random_state. Expected int or "
+                f"np.random.RandomState but got {type(random_state)}."
+            )
+
     # Extract information from the decomposition dictionary
     probs = get_probabilities(ideal_operation, decomposition_dict)
     one_norm = get_one_norm(ideal_operation, decomposition_dict)
 
     # Sample an index from the distribution "probs"
-    idx = np.random.choice(list(range(len(probs))), p=probs)
+    idx = rng.choice(range(len(probs)), p=probs)
 
     # Get the coefficient and the implementable sequence associated to "idx"
     coeff, imp_seq = decomposition_dict[ideal_operation][idx]
@@ -61,7 +77,9 @@ def sample_sequence(
 
 
 def sample_circuit(
-    ideal_circuit: Circuit, decomposition_dict: DecompositionDict
+    ideal_circuit: Circuit,
+    decomposition_dict: DecompositionDict,
+    random_state: Optional[Union[int, np.random.RandomState]] = None
 ) -> Tuple[Circuit, int, float]:
     """Samples an implementable circuit according from the PEC decomposition
     of the input ideal circuit. Moreover it also returns the "sign" and "norm"
@@ -73,6 +91,7 @@ def sample_circuit(
         decomposition_dict: The decomposition dictionary containing the quasi-
             probability representation of the ideal operations (those
             which are part of "ideal_circuit").
+        random_state: Seed for sampling.
 
     Returns:
         imp_circuit: The sampled implementable circuit.
@@ -90,7 +109,7 @@ def sample_circuit(
     for ideal_operation in ideal_circuit.all_operations():
         # Sample an imp. sequence from the decomp. of ideal_operation
         imp_seq, loc_sign, loc_norm = sample_sequence(
-            ideal_operation, decomposition_dict
+            ideal_operation, decomposition_dict, random_state
         )
         sign *= loc_sign
         norm *= loc_norm

--- a/mitiq/pec/sampling.py
+++ b/mitiq/pec/sampling.py
@@ -79,7 +79,7 @@ def sample_sequence(
 def sample_circuit(
     ideal_circuit: Circuit,
     decomposition_dict: DecompositionDict,
-    random_state: Optional[Union[int, np.random.RandomState]] = None
+    random_state: Optional[Union[int, np.random.RandomState]] = None,
 ) -> Tuple[Circuit, int, float]:
     """Samples an implementable circuit according from the PEC decomposition
     of the input ideal circuit. Moreover it also returns the "sign" and "norm"

--- a/mitiq/pec/sampling.py
+++ b/mitiq/pec/sampling.py
@@ -50,7 +50,7 @@ def sample_sequence(
         sign: The sign associated to sampled sequence.
         norm: The one norm of the decomposition coefficients.
     """
-    if not random_state:
+    if random_state is None:
         rng = np.random
     else:
         if isinstance(random_state, int):

--- a/mitiq/pec/tests/test_pec.py
+++ b/mitiq/pec/tests/test_pec.py
@@ -45,11 +45,7 @@ def executor(circuit: Circuit) -> float:
     else:
         raise ValueError("The input must be a circuit with 1 or 2 qubits.")
 
-    return noisy_simulation(
-        circuit,
-        BASE_NOISE,
-        obs,
-    )
+    return noisy_simulation(circuit, BASE_NOISE, obs,)
 
 
 # Simple identity 1-qubit circuit for testing
@@ -58,11 +54,7 @@ oneq_circ = Circuit(Z.on(q), Z.on(q))
 
 # Simple identity 2-qubit circuit for testing
 qreg = LineQubit.range(2)
-twoq_circ = Circuit(
-    Y.on(qreg[1]),
-    CNOT.on(*qreg),
-    Y.on(qreg[1]),
-)
+twoq_circ = Circuit(Y.on(qreg[1]), CNOT.on(*qreg), Y.on(qreg[1]),)
 
 
 @mark.parametrize("circuit", [oneq_circ, twoq_circ])
@@ -111,7 +103,7 @@ def test_execute_with_pec_with_different_samples(circuit: Circuit, seed):
             executor,
             decomposition_dict=DECO_DICT,
             num_samples=100,
-            random_state=seed
+            random_state=seed,
         )
         errors_more_samples.append(abs(mitigated - 1.0))
 

--- a/mitiq/pec/tests/test_pec.py
+++ b/mitiq/pec/tests/test_pec.py
@@ -90,7 +90,8 @@ def test_execute_with_pec_one_qubit(
 
 
 @mark.parametrize("circuit", [oneq_circ, twoq_circ])
-def test_execute_with_pec_with_different_samples(circuit: Circuit):
+@mark.parametrize("seed", (1, 2, 3))
+def test_execute_with_pec_with_different_samples(circuit: Circuit, seed):
     """Tests that, on average, the error decreases as the number of samples is
     increased.
     """
@@ -98,11 +99,19 @@ def test_execute_with_pec_with_different_samples(circuit: Circuit):
     errors_more_samples = []
     for _ in range(10):
         mitigated = execute_with_pec(
-            circuit, executor, decomposition_dict=DECO_DICT, num_samples=10
+            circuit,
+            executor,
+            decomposition_dict=DECO_DICT,
+            num_samples=10,
+            random_state=seed,
         )
         errors_few_samples.append(abs(mitigated - 1.0))
         mitigated = execute_with_pec(
-            circuit, executor, decomposition_dict=DECO_DICT, num_samples=100
+            circuit,
+            executor,
+            decomposition_dict=DECO_DICT,
+            num_samples=100,
+            random_state=seed
         )
         errors_more_samples.append(abs(mitigated - 1.0))
 

--- a/mitiq/pec/tests/test_pec_sampling.py
+++ b/mitiq/pec/tests/test_pec_sampling.py
@@ -46,10 +46,7 @@ NOISELESS_DECO_DICT = _simple_pauli_deco_dict(0)
 
 # Simple 2-qubit circuit
 qreg = LineQubit.range(2)
-twoq_circ = Circuit(
-    X.on(qreg[0]),
-    CNOT.on(*qreg),
-)
+twoq_circ = Circuit(X.on(qreg[0]), CNOT.on(*qreg),)
 
 
 @mark.parametrize("gate", [X, Y, Z, CNOT])

--- a/mitiq/pec/tests/test_pec_sampling.py
+++ b/mitiq/pec/tests/test_pec_sampling.py
@@ -30,6 +30,7 @@ from cirq import (
     depolarize,
 )
 
+from mitiq.utils import _equal
 from mitiq.pec.utils import (
     _simple_pauli_deco_dict,
     DecompositionDict,
@@ -62,11 +63,56 @@ def test_sample_sequence_types(gate: Gate):
         assert norm > 1
 
 
+@mark.parametrize("seed", (1, 2, 3, 5))
+@mark.parametrize("seed_type", (int, np.random.RandomState))
+@mark.parametrize("op", [X(qreg[0]), Y(qreg[0]), Z(qreg[0]), CNOT(*qreg)])
+def test_sample_sequence_random_state(seed, seed_type, op):
+    decomposition = _simple_pauli_deco_dict(0.5)
+
+    if isinstance(seed_type, np.random.RandomState):
+        seed = np.random.RandomState(seed)
+    sequence, sign, norm = sample_sequence(
+        op, decomposition, random_state=seed
+    )
+
+    for _ in range(20):
+        if isinstance(seed_type, np.random.RandomState):
+            seed = np.random.RandomState(seed)
+        new_sequence, new_sign, new_norm = sample_sequence(
+            op, decomposition, random_state=seed
+        )
+        assert new_sequence == sequence
+        assert new_sign == sign
+        assert np.isclose(new_norm, norm)
+
+
 def test_sample_circuit_types():
     imp_circuit, sign, norm = sample_circuit(twoq_circ, DECO_DICT)
     assert isinstance(imp_circuit, Circuit)
     assert sign in {1, -1}
     assert norm > 1
+
+
+@mark.parametrize("seed", (1, 2, 3, 5))
+@mark.parametrize("seed_type", (int, np.random.RandomState))
+def test_sample_circuit_random_state(seed, seed_type):
+    decomposition = _simple_pauli_deco_dict(0.5)
+
+    if isinstance(seed_type, np.random.RandomState):
+        seed = np.random.RandomState(seed)
+    circuit, sign, norm = sample_circuit(
+        twoq_circ, decomposition, random_state=seed
+    )
+
+    for _ in range(20):
+        if isinstance(seed_type, np.random.RandomState):
+            seed = np.random.RandomState(seed)
+        new_circuit, new_sign, new_norm = sample_circuit(
+            twoq_circ, decomposition, random_state=seed
+        )
+        assert _equal(new_circuit, circuit)
+        assert new_sign == sign
+        assert np.isclose(new_norm, norm)
 
 
 def test_sample_circuit_types_trivial():

--- a/mitiq/pec/tests/test_pec_sampling.py
+++ b/mitiq/pec/tests/test_pec_sampling.py
@@ -62,7 +62,9 @@ def test_sample_sequence_types(gate: Gate):
 
 @pytest.mark.parametrize("seed", (1, 2, 3, 5))
 @pytest.mark.parametrize("seed_type", (int, np.random.RandomState))
-@pytest.mark.parametrize("op", [X(qreg[0]), Y(qreg[0]), Z(qreg[0]), CNOT(*qreg)])
+@pytest.mark.parametrize(
+    "op", [X(qreg[0]), Y(qreg[0]), Z(qreg[0]), CNOT(*qreg)]
+)
 def test_sample_sequence_random_state(seed, seed_type, op):
     decomposition = _simple_pauli_deco_dict(0.5)
 

--- a/mitiq/pec/tests/test_pec_sampling.py
+++ b/mitiq/pec/tests/test_pec_sampling.py
@@ -15,7 +15,7 @@
 
 """Tests related to mitiq.pec.sampling functions."""
 
-from pytest import mark
+import pytest
 import numpy as np
 
 from cirq import (
@@ -49,7 +49,7 @@ qreg = LineQubit.range(2)
 twoq_circ = Circuit(X.on(qreg[0]), CNOT.on(*qreg),)
 
 
-@mark.parametrize("gate", [X, Y, Z, CNOT])
+@pytest.mark.parametrize("gate", [X, Y, Z, CNOT])
 def test_sample_sequence_types(gate: Gate):
     num_qubits = gate.num_qubits()
     qreg = LineQubit.range(num_qubits)
@@ -60,9 +60,9 @@ def test_sample_sequence_types(gate: Gate):
         assert norm > 1
 
 
-@mark.parametrize("seed", (1, 2, 3, 5))
-@mark.parametrize("seed_type", (int, np.random.RandomState))
-@mark.parametrize("op", [X(qreg[0]), Y(qreg[0]), Z(qreg[0]), CNOT(*qreg)])
+@pytest.mark.parametrize("seed", (1, 2, 3, 5))
+@pytest.mark.parametrize("seed_type", (int, np.random.RandomState))
+@pytest.mark.parametrize("op", [X(qreg[0]), Y(qreg[0]), Z(qreg[0]), CNOT(*qreg)])
 def test_sample_sequence_random_state(seed, seed_type, op):
     decomposition = _simple_pauli_deco_dict(0.5)
 
@@ -90,8 +90,8 @@ def test_sample_circuit_types():
     assert norm > 1
 
 
-@mark.parametrize("seed", (1, 2, 3, 5))
-@mark.parametrize("seed_type", (int, np.random.RandomState))
+@pytest.mark.parametrize("seed", (1, 2, 3, 5))
+@pytest.mark.parametrize("seed_type", (int, np.random.RandomState))
 def test_sample_circuit_random_state(seed, seed_type):
     decomposition = _simple_pauli_deco_dict(0.5)
 
@@ -112,6 +112,11 @@ def test_sample_circuit_random_state(seed, seed_type):
         assert np.isclose(new_norm, norm)
 
 
+def test_sample_sequence_bad_random_state():
+    with pytest.raises(ValueError, match="Bad type for random_state."):
+        sample_sequence(X(qreg[0]), DECO_DICT, random_state="")
+
+
 def test_sample_circuit_types_trivial():
     imp_circuit, sign, norm = sample_circuit(twoq_circ, NOISELESS_DECO_DICT)
     assert imp_circuit == twoq_circ
@@ -119,7 +124,7 @@ def test_sample_circuit_types_trivial():
     assert np.isclose(norm, 1)
 
 
-@mark.parametrize("gate", [Y, CNOT])
+@pytest.mark.parametrize("gate", [Y, CNOT])
 def test_sample_sequence_choi(gate: Gate):
     """Tests the sample_sequence by comparing the exact Choi matrices."""
     qreg = LineQubit.range(gate.num_qubits())
@@ -142,7 +147,7 @@ def test_sample_sequence_choi(gate: Gate):
     assert np.allclose(ideal_choi, choi_pec_estimate, atol=0.05)
 
 
-@mark.parametrize("decomposition_dict", [DECO_DICT, DECO_DICT_SIMP])
+@pytest.mark.parametrize("decomposition_dict", [DECO_DICT, DECO_DICT_SIMP])
 def test_sample_circuit_choi(decomposition_dict: DecompositionDict):
     """Tests the sample_circuit by comparing the exact Choi matrices."""
     ideal_choi = _circuit_to_choi(twoq_circ)

